R/formatter.r

```diff
@@ -1,3 +1,84 @@
+#' Number formatter: a generic formatter for numbers
+#'
+#' @return `number_format` returns a function with single parameter
+#' `x`, a numeric vector, that returns a character vector
+#' @param x a numeric vector to format
+#' @param accuracy number to round to, `NULL` for automatic guess
+#' @param scale a scaling factor: `x` will be multiply by `scale` before
+#' formating (useful if the underlying data is on another scale,
+#' e.g. for computing percentages or thousands)
+#' @param prefix,suffix symbols to display before and after value
+#' @param big.mark character used between every 3 digits to separate thousands,
+#' a comma by default or a space if comma is already used for `decimal.mark`
+#' @param decimal.mark the character to be used to indicate the numeric
+#' decimal point
+#' @param trim logical, if `FALSE`, values are right-justified to a common
+#' width (see [base::format()])
+#' @param ... other arguments passed on to [base::format()]
+#' @export
+#' @examples
+#' v <- c(12.3, 4, 12345.789, 0.0002)
+#' number(v)
+#' number(v, big.mark = ",")
+#' number(v, accuracy = .001)
+#' number(v, accuracy = .001, decimal.mark = ",")
+#' number(v, accuracy = .5)
+#'
+#' my_format <- number_format(big.mark = "'", decimal.mark = ",")
+#' my_format(v)
+#'
+#' # Per mille
+#' per_mille <- number_format(
+#'   scale = 1000,
+#'   suffix = "\u2030",
+#'   accuracy = .1
+#' )
+#' per_mille(v)
+number_format <- function(accuracy = 1, scale = 1, prefix = "",
+                          suffix = "", big.mark = NULL, decimal.mark = ".",
+                          trim = TRUE, ...) {
+  function(x) number(
+      x,
+      accuracy = accuracy,
+      scale = scale,
+      prefix = prefix,
+      suffix = suffix,
+      big.mark = big.mark,
+      decimal.mark = decimal.mark,
+      trim = trim,
+      ...
+    )
+}
+
+#' @export
+#' @rdname number_format
+number <- function(x, accuracy = 1, scale = 1, prefix = "",
+                   suffix = "", big.mark = NULL, decimal.mark = ".",
+                   trim = TRUE, ...) {
+  if (length(x) == 0) return(character())
+  accuracy <- accuracy %||% precision(x)
+  x <- round_any(x, accuracy / scale)
+  nsmall <- -floor(log10(accuracy))
+  nsmall <- min(max(nsmall, 0), 20)
+
+  if (is.null(big.mark)) {
```

Comment:
Hmmmm, now this logic seems like it's the wrong way around. Maybe it's `decimal.mark` that should be `NULL`?

This is making me feel that you're correct that the default big.mark should be `" "`

Reply:
It's not so easy to find an appropriate behaviour if we set `decimal.mark` to `NULL`.

Your previous proposal was:

```r
if (is.null(decimal.mark)) {
  decimal.mark = if (identical(big.mark, ",")) "." else ","
}
```

But this behaviour is probably not the best. If I need to present results in an international format, I have to specify `big.mark = " "` but I want to keep a dot as a decimal separator.

If I want to present results in French, the first thing I want to set is the `decimal.mark` as a comma, and the change of `big.mark` is a consequence of this previous change.

Therefore, there is no general rule to derive `decimal.mark` from `big.mark`.

It is true that using a space as the default thousands separator is more generic as it is working with different decimal marks.

In addition, for those in need to present results in an American style, there is still the `comma_format` that will be simply a short cut for `number_format` with a comma as the default decimal mark.

Reply:
Ok, lets go back to your original proposal with `big.mark = " "`

Reply:
done